sys/initrd.c

```diff
@@ -0,0 +1,266 @@
+ * inside `p2` directory. */
+static bool is_direct_descendant(const char *p1, const char *p2) {
+  while (*p1 && *p1 == *p2) {
+    p1++, p2++;
+  }
+
+  if (*p2)
+    return false;
+
+  /* Check whether p is valid filename (does not contain '/') */
+  p1++; /* skip trailing '/' */
+  return (strchr(p1, '/') == NULL);
+}
+
+/* Extract last component of the path. */
+static const char *basename(const char *path) {
+  char *name = strrchr(path, '/');
+  return name ? name + 1 : path;
+}
+
+static void initrd_build_tree_and_names() {
+  cpio_node_t *it_i, *it_j;
+
+  TAILQ_FOREACH (it_i, &initrd_head, c_list) {
+    TAILQ_FOREACH (it_j, &initrd_head, c_list) {
+      if (it_i == it_j)
+        continue;
+      if (is_direct_descendant(it_j->c_path, it_i->c_path))
+        TAILQ_INSERT_TAIL(&it_i->c_children, it_j, c_siblings);
+    }
+  }
+
+  cpio_node_t *it;
+  TAILQ_FOREACH (it, &initrd_head, c_list) {
+    it->c_name = basename(it->c_path);
+  }
+
+  /* Construct a node that represent the root of filesystem. */
+  root_node = cpio_node_alloc();
+  root_node->c_path = "";
+  TAILQ_FOREACH (it, &initrd_head, c_list) {
+    if (is_direct_descendant(it->c_path, "")) {
+      TAILQ_INSERT_TAIL(&root_node->c_children, it, c_siblings);
+    }
+  }
+}
+
+static int initrd_vnode_lookup(vnode_t *vdir, const char *name, vnode_t **res) {
+  cpio_node_t *it;
+  cpio_node_t *cn_dir = (cpio_node_t *)vdir->v_data;
+
+  TAILQ_FOREACH (it, &cn_dir->c_children, c_siblings) {
+    if (strcmp(name, it->c_name) == 0) {
+      vnodetype_t type = V_REG;
+      if (it->c_mode & C_ISDIR)
+        type = V_DIR;
+      *res = vnode_new(type, &initrd_ops);
+      (*res)->v_data = (void *)it;
+      vnode_ref(*res);
+      return 0;
+    }
+  }
+  return ENOENT;
+}
+
+static int initrd_vnode_read(vnode_t *v, uio_t *uio) {
+  cpio_node_t *cn = (cpio_node_t *)v->v_data;
+  int count = uio->uio_resid;
+  int error = uiomove(cn->c_data, cn->c_size, uio);
+
+  if (error < 0)
+    return -error;
+
+  return count - uio->uio_resid;
+}
+
+static int initrd_mount(mount_t *m) {
+  vnode_t *root = vnode_new(V_DIR, &initrd_ops);
+  root->v_data = (void *)root_node;
+  root->v_mount = m;
+  m->mnt_data = root;
+  return 0;
+}
+
+static int initrd_root(mount_t *m, vnode_t **v) {
+  *v = m->mnt_data;
+  vnode_ref(*v);
+  return 0;
+}
+
+static int initrd_init(vfsconf_t *vfc) {
+  vfs_domount(vfc, vfs_root_initrd_vnode);
+  return 0;
+}
+
+intptr_t ramdisk_get_start() {
+  char *s = kenv_get("rd_start");
+  if (s == NULL)
+    return 0;
+  int s_len = strlen(s);
+  return strtoul(s + s_len - 8, NULL, 16);
+}
+
+unsigned ramdisk_get_size() {
+  char *s = kenv_get("rd_size");
+  return s ? strtoul(s, NULL, 0) : 0;
+}
+
+void ramdisk_init() {
+  unsigned rd_size = ramdisk_get_size();
+
+  TAILQ_INIT(&initrd_head);
+  kmalloc_init(mp);
+  kmalloc_add_pages(mp, 2);
+
+  if (rd_size) {
+    initrd_ops.v_lookup = initrd_vnode_lookup;
+    initrd_ops.v_read = initrd_vnode_read;
```

Comment:
`v_readdir` and `v_getattr` are missing, but this PR is large enough to ignore this issue for now.